upstairs/src/io_limits.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2024 Oxide Computer Company
+use crate::{ClientData, ClientId, ClientMap};
+use std::sync::Arc;
+use tokio::sync::{
+    AcquireError, OwnedSemaphorePermit, Semaphore, TryAcquireError,
+};
+
+/// Per-client IO limits
+#[derive(Clone, Debug)]
+pub struct ClientIOLimits {
+    /// Semaphore to claim IO bytes on behalf of a job
+    io_bytes: Arc<Semaphore>,
+
+    /// Semaphore to claim individual IO jobs
+    jobs: Arc<Semaphore>,
+}
+
+impl ClientIOLimits {
+    /// Builds a new `ClientIOLimits` object with the given limits
+    pub fn new(max_jobs: usize, max_io_bytes: usize) -> Self {
+        ClientIOLimits {
+            io_bytes: Semaphore::new(max_io_bytes).into(),
+            jobs: Semaphore::new(max_jobs).into(),
+        }
+    }
+
+    /// Claims a certain number of bytes (and one job)
+    ///
+    /// This function waits until the given resources are available; as such, it
+    /// should not be called from the same task which is processing jobs (since
+    /// that could create a deadlock).
+    async fn claim(
+        &self,
+        bytes: u32,
+    ) -> Result<ClientIOLimitGuard, AcquireError> {
+        let io_bytes = self.io_bytes.clone().acquire_many_owned(bytes).await?;
+        let jobs = self.jobs.clone().acquire_owned().await?;
+        Ok(ClientIOLimitGuard { io_bytes, jobs })
+    }
+
+    /// Tries to claim a certain number of bytes (and one job)
+    pub fn try_claim(
+        &self,
+        bytes: u32,
+    ) -> Result<ClientIOLimitGuard, TryAcquireError> {
+        let io_bytes = self.io_bytes.clone().try_acquire_many_owned(bytes)?;
+        let jobs = self.jobs.clone().try_acquire_owned()?;
+        Ok(ClientIOLimitGuard { io_bytes, jobs })
+    }
```

Comment:
Coming back to these two, same suggestion: if there's an error acquiring permits for the jobs semaphore, the acquired permits for the io_bytes semaphore should be released